feat: add WebRTC (MediaMTX / rtsp-simple-server) webcam mode

slynn1324 opened the pull request:
Adds support for ws://... urls in the IP Camera webcam type.

See https://github.com/mainsail-crew/mainsail/issues/1317.

meteyou commented:
Thx for you your PR! The main problem with WebRTC is that there are so many different "handshakes" for WebRTC. It's not possible to have one webrtc mode to support all streamers.

I only would prefer to have a separate mode for this webcam style. It looks like the "wyze-bridge" is using the rtsp-simple-server in the background for the webrtc stream. So I think the name "WebRTC (rtsp-simple-server)" would be a good name for this mode.

The mode "IP-Webcam" is for generic IP-Webcams.

slynn1324 commented:
Fair enough... I didn't dig into all the details of the underlying protocols (just hacked together what worked from an example) but that makes sense.  I'll look at refactoring this out to a separate camera type and update this PR.  Thanks!

slynn1324 commented:
Updated as a new camera type.  I think I poked around and found the necessary references that the camera types need expanded.  Please take a look.

meteyou commented on src/components/webcams/WebrtcRTSPSimpleServer.vue:
```diff
@@ -0,0 +1,251 @@
+
+        if (this.webRtcWs) {
+            try {
+                this.webRtcWs.close()
```
this.webRtcWs?.close() should be enough here. i dont think you need a try...catch statement

meteyou commented on src/components/webcams/WebrtcRTSPSimpleServer.vue:
```diff
@@ -0,0 +1,251 @@
+        }
+        if (this.webRtcPc) {
+            try {
+                this.webRtcPc.close()
```
this.webRtcPs?.close() should be enough here. i don't think you need a try...catch statement

meteyou commented on src/components/webcams/WebrtcRTSPSimpleServer.vue:
```diff
@@ -0,0 +1,251 @@
+        }
+
+        this.webRtcRestartTimeout = window.setTimeout(() => {
+            this.webRtcRestartTimeout = null
```
pls clear the timeout in the webRtcStart function and set it there to null

meteyou commented on src/components/webcams/WebrtcRTSPSimpleServer.vue:
```diff
@@ -0,0 +1,251 @@
+        this.webRtcTerminate()
+    }
+
+    visibilityChanged(isVisible: boolean) {
```
i would not use the visibilty check for webRTC, because webRTC needs too long to reconnect and it also need only a small bandwidth to work.

meteyou commented on src/components/webcams/WebrtcRTSPSimpleServer.vue:
```diff
@@ -0,0 +1,251 @@
+<style scoped>
```
pls move the style tag to the end of the vue file (i'm refactoring every file, when i see this "wrong" sort)

meteyou reviewed (commented):
in general, this already looks very good. i added a few comments, but after that it should fit. i will try to install this streamer to be able to test this.

slynn1324 commented:
See how that is.  I reworked the start/stop logic to track when the panel is actively collapsed vs. just scrolled out of view.  I also had to make a small change to TheSettingsMenu.vue so that the component would be destroyed if you click on the 'x' to close to the settings dialog without either cancelling or updating the camera in the dialog first.  I think this strikes a good balance between not reconnecting too frequently yet only keeping the streaming connection open if it stands to be displayed.   

I think I addressed all of the other comments.  

Thanks.

meteyou commented:
I cleaned the file and added TS types as well as possible. Yesterday also, the project rtsp-simple-server was renamed to MediaMTX. So I also renamed it in this PR. I tested it with a newly installed MediaMTX with a raspberry webcam. Please double-check if it also works in your case, and then this PR will be merge-ready.

slynn1324 commented:
Apologies for the delay in getting back here.  I pulled the updates and installed them and it seems to work fine with my setup here and the wyze-bridge stream.  

I did notice 2 things however that were working better (in my opinion) prior to your updates:

1) The web socket connection would not be opened when the panel was not expanded.  The way that the panels seem to work on the dashboard (presumably due to animations), the components inside are mounted even when they are not visible.  I do not think that the additional load on the backing servers (or the bandwidth) should be used to open the stream if it is not visible.

2) The controls attribute was removed from the <video> element.  The controls allow to stop and start the stream - again for bandwidth purposes if desired.  What is even more useful is on mobile (at least iOS devices), including the controls allows for opening the video in full screen natively.  While not having any controls is probably a bit more consistent with some of the other video formats - I'd encourage to include them here where it's natively supported.

I typically have mainsail up on 3 devices simultaneously:  A monitor by the printer -- no camera needed there. I keep the Webcam panel minimized), my main PC in my office where I do most of my work, but monitor the printer from another room in my house, and then checking in from iPhone or iPad devices when I'm not in one of those places.  

Of the two issues, I think #1 should be looked at more closely and corrected for to be kind to users bandwidth and device needs, and #2 is more of a preference.

meteyou commented:
@slynn1324 thx for your high-level feedback! You are right about point 1. I saw it as "unimportant" when refactoring and wanted to remove unnecessary complexity, but with your explanation of your workflow, this sounds very reasonable! Do you want to add this back, or should I do it?

I intentionally removed the controls because not all functions work and would only lead to "wrong bug reports". I did the same with the WebRTC (camera-streamer) client. But I wouldn't know how to hide specific controls, so I removed them all. We can gladly enable them again if you know how to show and hide them to ensure 100% compatibility.

slynn1324 commented:
If you want to add back #1 as you'd like it to be that'd be great.  It might be a bit before I'm able to get back into the code.  

As for the controls - I would doubt there is a way to use only some of the native controls.  On my iOS devices, it shows up with the native player controls which then allows full screen and pausing etc.  It also displays that it's a live stream.  Would this be something to consider an extra option in the webcam configuration panel about whether the native controls should be displayed or not?  I could really see different users wanting it different ways... depending on their preferences.

meteyou commented:
@slynn1324 i pushed a few changes to fix the closed panel behavior. it was included in the code, but with a bug.

Please test the build again.

slynn1324 commented:
Tested here and seems to be working fine.  I would think this is good to merge whenever you're ready.

meteyou commented on src/components/panels/WebcamPanel.vue:
```diff
@@ -63,6 +63,9 @@
                         <webcam-webrtc-camerastreamer :cam-settings="currentCam" />
                     </template>
+                    <template v-else-if="currentCam.service === 'webrtc-mediamtx'">
+                        <webcam-webrtc-mediamtx :cam-settings="currentCam" :collapsible="$route.fullPath !== '/cam'" />
```
```suggestion
                        <webcam-webrtc-mediamtx :cam-settings="currentCam" />
```

this variable is not used in the component anymore

meteyou commented on src/components/webcams/WebrtcMediaMTX.vue:
```diff
@@ -0,0 +1,238 @@
+@Component
+export default class WebrtcRTSPSimpleServer extends Mixins(BaseMixin) {
+    @Prop({ required: true }) camSettings: any
+    @Prop({ default: false }) collapsible: boolean | undefined
```
```suggestion

```

this prop is not used in this component anymore

meteyou commented on src/components/TheSettingsMenu.vue:
```diff
@@ -55,7 +55,10 @@
-                            <component :is="'settings-' + activeTab + '-tab'" @scrollToTop="scrollToTop"></component>
+                            <component
+                                :is="'settings-' + activeTab + '-tab'"
+                                v-if="showSettings"
```
Why do you add here `v-if="showSettings"`?

meteyou commented:
Ok. I reviewed the code in detail and have found some little refactorings and one question. But the rest looks good to me. We also have to update this branch before merging it because other webcam clients were added to the developer branch, and there are merge conflicts right now.

meteyou commented:
@slynn1324 any updates here?

meteyou commented:
@slynn1324 can you share any information how to setup this webcam in wyze-bridge and which url you need to use in mainsail?

Naninani commented:
@meteyou @slynn1324 Could you share instructions on how to setup?
